### ISSUE TITLE
Consider changing `issuer-url` to `oidc-issuer-location`

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-boot#14222]
The properties for configuring an issuer in boot are

`security.oauth2.resourceserver.jwt.issuer-url` and 
`security.oauth2.client.provider.{registrationId}.issuer-url`.

This is nice because the naming is consistent and it is clear based on the naming what they are configuring (a url).

Because this property is for the expressed purpose of configuring via the [OIDC Provider Configuration specification](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig), it's valuable to further elaborate on what they are configuring by including additional spec language into the name. One reference in the spec reads:

> Using the Issuer location discovered as described in Section 2 or by other means, the OpenID Provider's configuration information can be retrieved.

In fact, whenever the issuer value is referred to in the spec, the language is "Issuer location".

A quick google of "oidc issuer location" (in an incognito window) shows the related specification as the first result, which will assist the user in understanding the meaning of the property.

While it is true that "location" is more general than "url", if I open an incognito window and google for "oidc issuer url", the spec does not come back as the first link but instead is about half-way down the page.

Including `oidc` as a prefix to the name makes so that Spring Boot does not need to guess the user's intent in setting the property. It also makes it clear to the user that, by providing this property, they are configuring the application using that specification.

There is a separate use case to consider as well, and that is when the user just wants to validate the issuer and does _not_ want to do discovery. First, we want to be clear that this is not the purpose of this property, and second, we don't want to risk trying to conflate two separate use cases, making Spring Boot try and guess the user's intent (e.g. does the user want to do discovery or just issuer validation?).

If the property remains `issuer-url`, there is the risk that the user will misinterpret the intent of the property. There is also the concern of what will be the property name for just doing issuer validation, should we add that down the road.

**Comment by philwebb** [comment]
I'm not a big fan of acronyms in properties so`oidc-issuer-location` doesn't sit well with me. I guess we could name it `openid-connect-issuer-location`, but that's getting a bit wordy.

Perhaps we can keep the `issuer-url` name (which I quite like) and use the property meta-data to provide more details on the purpose?

**Comment by vpavic** [comment]
With the emergence of OAuth 2.0 Authorization Server Metadata specification (i.e. [RFC 8414](https://tools.ietf.org/html/rfc8414)) I'd strongly suggest against using anything OpenID Connect specific in the metadata related properties (such as this one) since going forward metadata format specified in this RFC should be preferred to [OpenID Connect Discovery 1.0](https://openid.net/specs/openid-connect-discovery-1_0.html) - see RFC 8414 [introduction](https://tools.ietf.org/html/rfc8414#section-1) and [compatiblity notes](https://tools.ietf.org/html/rfc8414#section-5) sections.

**Comment by mbhave** [comment]
Just to clarify, the current property is `issuerUri` and not `issuerUrl`. I think that is consistent with the rest of the properties we have for other OAuth2/Open ID connect endpoints. If there is any clarification that needs to be added about discovery vs validation we can add that in the documentation/metadata. 

Also, going by what @vpavic said, tying `oidc` or `open-id-connect` to the `issuer-uri` would prevent us from using the same property for the OAuth 2.0 Authorization Server Metadata spec.

My vote is for not changing the property and updating the documentation if necessary.

**Comment by snicoll** [comment]
I don't like acronyms in property names either. Clarifying the purpose of the key is exactly what the metadata and the documentation are for. Same vote as @mbhave.

**Comment by jzheaux** [comment]
> With the emergence of OAuth 2.0 Authorization Server Metadata specification (i.e. RFC 8414) I'd strongly suggest against using anything OpenID Connect specific in the metadata related properties

That is a good point, @vpavic, but then I ask, how will boot know which endpoint to address? For example, if the property is simply named `issuer-uri` (thank you for the correction, @mbhave), then I have three questions now:

1. Does the user want to use this for discovery?
2. If so, should I hit the oidc endpoint or the oauth2 endpoint?
3. If not, should I use this for issuer verification?

I think that this property name should clearly indicate that it is at least for the purpose of discovery. In the end, I suppose the code could attempt to hit both the oauth2 and oidc metadata endpoints in some defined order if that were the case. 

The property ought to clearly indicate the intent, e.g. "I want you to use the oidc metadata endpoint, using this property, to configure the rest", and I believe `oidc-issuer-location` does that. If the user wanted to say "I want you to use the oauth2 metadata endpoint, using this property, to configure the rest", we would have a different property for that.

> I guess we could name it openid-connect-issuer-location, but that's getting a bit wordy.

@philwebb I don't mind a wordier property, especially if IDEs autocomplete it for me anyway.

**Comment by vpavic** [comment]
> how will boot know which endpoint to address?

IMO this is the domain concern and as such it should be handled in Security, not Boot. I'd like to again point to [compatiblity notes](https://tools.ietf.org/html/rfc8414#section-5) section of RFC 8414 which addresses some of your questions:

> Going forward, OAuth authorization server metadata locations should use the transformation defined in this specification. However, when deployed in legacy environments in which the OpenID Connect Discovery 1.0 transformation is already used, it may be necessary during a transition period to publish metadata for issuer identifiers containing a path component at both locations. **During this transition period, applications should first apply the transformation defined in this specification and attempt to retrieve the authorization server metadata from the resulting location; only if the retrieval from that location fails should they fall back to attempting to retrieve it from the alternate location obtained using the transformation defined by OpenID Connect Discovery 1.0.** This backwards-compatible behavior should only be necessary when the well-known URI suffix employed by the application is "openid-configuration".

> I think that this property name should clearly indicate that it is at least for the purpose of discovery.

If the property is documented that it serves the purpose of configuring from provider metadata that should be enough, right?
